### PR TITLE
AP_HAL: clean up UART driver storage/access and legacy ordering

### PR DESCRIPTION
**Pull request by tpwrules**
Eliminates UB and saves 108 bytes on CubeOrangePlus. See commits for details.

No functional change.  I did test this on a real board and confirm using a GPS that the port ordering is the same.

Remaining shadows of the legacy ordering:
* TCP port numbers on SITL (changing will break users)
* ESP32 UART indices (can't test, don't want to fiddle with)
* SITL and Linux command line arguments (removing will break users, deprecated with warning)

### CODE REVIEW

File: libraries/AP_HAL/HAL.h
Comment (peterbarker) [inline libraries/AP_HAL/HAL.h:161]:
```suggestion
    if (sernum >= ARRAY_SIZE(uart_array)) {
```

**Review by peterbarker — APPROVED** [review]
LGTM.

It's progress (removing `uartX` as members is good), but there's obviously still more to do here.

Next step would probably be to push that mapping back down into each of the HALs, so the HALs can push the devices into here in a 1-1 order.

**Review by tridge — CHANGES_REQUESTED** [review]
I want to get rid of the old mapping completely, just needs some work in HAL_SITL, HAL_Linux etc

**Comment by tpwrules** [comment]
> I want to get rid of the old mapping completely, just needs some work in HAL_SITL, HAL_Linux etc

I looked into this but the old mapping still is user-visible as command line options with those. Is it finally time for those options to be removed? That could break people's scripts.

**Comment by rmackay9** [comment]
@tridge how do you feel about merging this now and then do the AP_HAL changes in a follow-up PR?  Just thinking of scope creep and helping devs cut their teeth on this smaller change.

**Comment by tpwrules** [comment]
I spent some cathartic time with Ctrl+F and I think I've managed to eliminate virtually every trace the old mapping. A couple things remain which might break users if removed. It shouldn't be much effort to remove them for good if desired.

**Comment by tridge** [comment]
@tpwrules wow, this is great! I've been putting this off for ages, and you did it exactly the way I wanted it done. I think you deserve for medal for taking this on.
We just need some testing of this on ChibiOS and SITL at least before merging. I can't spot any errors, but it is such a tricky conversion I think we need to do some careful testing

**Comment by tpwrules** [comment]
Thank you very much for the kind words. Glad to be able to help.

I did test one serial port with a serial GPS on my own CubeOrange to make sure that ordering did not change. SITL should be covered by the CI. I did also launch and fly around a bit a SITL system on my computer using MAVProxy and `sim_vehicle.py` but was not particularly exhaustive.

**Comment by tpwrules** [comment]
Warnings have been added. Should be ready to merge once you are happy with ChibiOS testing.

**Comment by tpwrules** [comment]
Final rebase, mostly to clean up some super minor things and make CI green.

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
@tpwrules I tested all uarts on cubeorange with no issues

**Pull request merged**
